lib/internal/quic/core.js

```diff
@@ -0,0 +1,3341 @@
+    // as we need to wait until ngtcp2 recognizes the stream as
+    // having been closed to be destroyed.
+
+    // Do nothing if we've already been destroyed
+    if (this.destroyed || this.#closed)
+      return;
+
+    if (this.pending)
+      return this.once('ready', () => this[kClose](family, code));
+
+    this.#closed = true;
+
+    this.#aborted = this.readable || this.writable;
+
+    // Trigger scheduling of the RESET_STREAM and STOP_SENDING frames
+    // as appropriate. Notify ngtcp2 that the stream is to be shutdown.
+    // Once sent, the stream will be closed and destroyed as soon as
+    // the shutdown is acknowledged by the peer.
+    this[kHandle].resetStream(code, family);
+
+    // Close down the readable side of the stream
+    if (this.readable) {
+      this.push(null);
+      this.read();
+    }
+
+    // It is important to call shutdown on the handle before shutting
+    // down the writable side of the stream in order to prevent an
+    // empty STREAM frame with fin set to be sent to the peer.
+    if (this.writable)
+      this.end();
+  }
+
+  [kAfterAsyncWrite]({ bytes }) {
+    // TODO(@jasnell): Implement this
+  }
+
+  [kInspect]() {
+    const direction = this.bidirectional ? 'bidirectional' : 'unidirectional';
+    const initiated = this.serverInitiated ? 'server' : 'client';
+    const obj = {
+      id: this.#id,
+      direction,
+      initiated,
+      writableState: this._writableState,
+      readableState: this._readableState,
+      stats: {
+        dataRate: this.dataRateHistogram,
+        dataSize: this.dataSizeHistogram,
+        dataAck: this.dataAckHistogram,
+      }
+    };
+    return `QuicStream ${util.format(obj)}`;
+  }
+
+  [kTrackWriteState](stream, bytes) {
+    // TODO(@jasnell): Not yet sure what we want to do with these
+    // this.#writeQueueSize += bytes;
+    // this.#writeQueueSize += bytes;
+    // this[kHandle].chunksSentSinceLastWrite = 0;
+  }
+
+  [kUpdateTimer]() {
+    // TODO(@jasnell): Implement this later
+  }
+
+  get pending() {
+    // The id is set in the kSetHandle function
+    return this.#id === undefined;
+  }
+
+  get aborted() {
+    return this.#aborted;
+  }
+
+  get serverInitiated() {
+    return !!(this.#id & 0b01);
+  }
+
+  get clientInitiated() {
+    return !this.serverInitiated;
+  }
+
+  get unidirectional() {
+    return !!(this.#id & 0b10);
+  }
+
+  get bidirectional() {
+    return !this.unidirectional;
+  }
+
+  #writeGeneric = function(writev, data, encoding, cb) {
+    if (this.destroyed)
+      return;  // TODO(addaleax): Can this happen?
+
+    // The stream should be corked while still pending
+    // but just in case uncork
+    // was called early, defer the actual write until the
+    // ready event is emitted.
+    if (this.pending) {
+      return this.once('ready', () => {
+        this.#writeGeneric(writev, data, encoding, cb);
+      });
+    }
+
+    this[kUpdateTimer]();
+    const req = (writev) ?
```

Comment:
redudant `()`

Reply:
Can address in subsequent pull request. The javascript side of this is likely to see a complete overhaul anyway.